src/shogun/classifier/Perceptron.cpp

```diff
@@ -78,44 +61,34 @@ bool CPerceptron::train_machine(CFeatures* data)
 		w.set_const(1.0 / num_feat);
 		bias=0;
 	}
-	else
-	{
-		w = get_w();
-	}
-	auto pb = SG_PROGRESS(range(max_iter));
-	//loop till we either get everything classified right or reach max_iter
-	while (!converged && iter < max_iter)
+}
+
+void CPerceptron::iteration()
+{
+	bool converged = true;
```

Comment:
could we set this to false up here and then set it to true if it is converged rather than the other way around?
Why does this even have to be a local variable? Can't we use the member directly? (and modify it if needed only)

Reply:
we can do directly with member but i think we talked about updating members in the end of the iteration ?

Reply:
The converged flag can be set to true by the iteration method. So there is no need to have a variable. Just set it to true when it is appropriate, and leave it unchanged otherwise